rclcpp/include/rclcpp/publisher.hpp

```diff
@@ -222,105 +73,72 @@ class Publisher : public PublisherBase
+  publish(std::unique_ptr<MessageT, MessageDeleter> msg)
   {
-    bool inter_process_subscriptions_exist =
-      get_subscription_count() > get_intra_process_subscription_count();
-    if (!intra_process_is_enabled_ || inter_process_subscriptions_exist) {
+    if (!intra_process_is_enabled_) {
       this->do_inter_process_publish(msg.get());
+      return;
     }
-    if (store_intra_process_message_) {
-      // Take the pointer from the unique_msg, release it and pass as a void *
-      // to the ipm. The ipm should then capture it again as a unique_ptr of
-      // the correct type.
-      // TODO(wjwwood):
-      //   investigate how to transfer the custom deleter (if there is one)
-      //   from the incoming unique_ptr through to the ipm's unique_ptr.
-      //   See: http://stackoverflow.com/questions/11002641/dynamic-casting-for-unique-ptr
-      MessageT * msg_ptr = msg.get();
-      msg.release();
-      uint64_t message_seq =
-        store_intra_process_message_(intra_process_publisher_id_, msg_ptr, typeid(MessageT));
-      rcl_interfaces::msg::IntraProcessMessage ipm;
-      ipm.publisher_id = intra_process_publisher_id_;
-      ipm.message_sequence = message_seq;
-      auto status = rcl_publish(&intra_process_publisher_handle_, &ipm);
-      if (RCL_RET_PUBLISHER_INVALID == status) {
-        rcl_reset_error();  // next call will reset error message if not context
-        if (rcl_publisher_is_valid_except_context(&intra_process_publisher_handle_)) {
-          rcl_context_t * context = rcl_publisher_get_context(&intra_process_publisher_handle_);
-          if (nullptr != context && !rcl_context_is_valid(context)) {
-            // publisher is invalid due to context being shutdown
-            return;
-          }
-        }
-      }
-      if (RCL_RET_OK != status) {
-        rclcpp::exceptions::throw_from_rcl_error(status, "failed to publish intra process message");
-      }
-    } else {
-      // Always destroy the message, even if we don't consume it, for consistency.
-      msg.reset();
-    }
-  }
-
-  virtual void
-  publish(const std::shared_ptr<MessageT> & msg)
-  {
-    // Avoid allocating when not using intra process.
-    if (!store_intra_process_message_) {
-      // In this case we're not using intra process.
-      return this->do_inter_process_publish(msg.get());
+    std::unique_ptr<MessageT, MessageDeleter> msg_copy;
+    // If an interprocess subscription exist, then the unique_ptr is promoted
+    // to a shared_ptr and published.
+    // This allows doing the intraprocess publish first and then doing the
+    // interprocess publish, resulting in lower publish-to-subscribe latency.
+    // It's not possible to do that with an unique_ptr,
+    // as do_intra_process_publish takes the ownership of the message.
+    if (get_subscription_count() > get_intra_process_subscription_count()) {
+      MessageSharedPtr shared_msg =
+        MessageSharedPtr(std::move(msg));
+      return this->publish(shared_msg);
```

Comment:
If you have both intraprocess and interprocess subscriptions, the idea here is to upgrade the `unique_ptr<T>` to a `shared_ptr<const T>`, then store the shared pointer in the `MRB` and then do the inter process publish. As I'm going to modify the `shared_ptr` publish method now, I will do that explicitly here.

Another option is possible: Do first interprocess publish and then store the `unique_ptr` in the `MRB`. That order is a little worst for publish-to-subscribe latency, although in some cases could end in avoiding one extra copy, e.g.: one `unique_ptr` subscription, one copy vs zero copy.

This is really a detail, I'm not completely sure of which decision is better. I will run a performance test comparing the last example I mentioned and then publish the results here.
It could be easily modified later without any API change, so modifying it again in a follow-up PR is not a big deal.

Reply:
But the whole point is that the publish being called here is unsafe. It may be used safely here, but the user can use it unsafely. If you want to upgrade the unique ptr to a shared ptr, that's fine, but you need to take an internal path to stored the newly created shared ptr in the MRB and then continue the logic.

For a `publish(shared_ptr<T>)` like signature to be safe, it needs to copy the referenced data, not just copy the shared ptr (taking shared ownership), otherwise other people who hold the shared pointer may mutate it after publish returns but before intra process finishes. So if you have to copy it, then this function could really be implemented as `return this->publish(*shared_msg);`.

Reply:
Yes, see the last commit. I already took an internal path.

Reply:
Yup, sorry still catching up :)